File: scripts/cleanup_db.py
import sqlite3
import os
import time

db_path = r"D:\repo github\sintak_pt_buya_barokah\database_dev.sqlite"

def cleanup():
    if not os.path.exists(db_path):
        print(f"File tidak ditemukan: {db_path}")
        return

    start_size = os.path.getsize(db_path) / (1024 * 1024 * 1024)
    print(f"--- Memulai Pembersihan Database ---")
    print(f"Ukuran awal: {start_size:.2f} GB")

    try:
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()

        # 1. Hapus log lama (sisakan hanya 3 hari terakhir)
        print("1. Menghapus log aktivitas yang lebih tua dari 3 hari...")
        cursor.execute("DELETE FROM activity_logs WHERE created_at < date('now', '-3 days')")
        deleted_logs = cursor.rowcount
        print(f"   Berhasil menghapus {deleted_logs:,} baris log.")

        # 2. Opsional: Kosongkan tabel FTS content jika memang ingin lebih bersih 
        # (FTS akan otomatis terisi kembali saat data diakses/diupdate)
        # Namun untuk sekarang kita fokus ke activity_logs saja.

        conn.commit()

        # 3. VACUUM (Ini yang akan mengecilkan ukuran file)
        print("2. Menjalankan VACUUM (Mengatur ulang ruang penyimpanan)...")
        print("   Mohon tunggu, proses ini memakan waktu beberapa menit untuk file 8GB...")
        
        start_vacuum = time.now() if hasattr(time, 'now') else time.time()
        conn.execute("VACUUM")
        end_vacuum = time.now() if hasattr(time, 'now') else time.time()
        
        print(f"   VACUUM selesai dalam {int(end_vacuum - start_vacuum)} detik.")

        conn.close()

        end_size = os.path.getsize(db_path) / (1024 * 1024 * 1024)
        print(f"\n--- Pembersihan Selesai ---")
        print(f"Ukuran Akhir: {end_size:.2f} GB")
        print(f"Ruang yang dihemat: {(start_size - end_size) * 1024:.2f} MB")

    except sqlite3.Error as e:
        print(f"Terjadi kesalahan database: {e}")
    except Exception as e:
        print(f"Terjadi kesalahan: {e}")

if __name__ == "__main__":
    cleanup()

File: scripts/cleanup-db.ts
import { createClient } from '@libsql/client';
import path from 'path';
import fs from 'fs';

const DAYS_TO_KEEP = 7;

function getFileSizeInMB(filePath: string): number {
  if (!fs.existsSync(filePath)) return 0;
  const stats = fs.statSync(filePath);
  return stats.size / (1024 * 1024);
}

async function cleanupDatabase(dbName: string) {
  const dbPath = path.join(process.cwd(), dbName);
  if (!fs.existsSync(dbPath)) {
    console.log(`[${dbName}] File tidak ditemukan, dilewati.`);
    return;
  }

  const initialSize = getFileSizeInMB(dbPath);
  console.log(`\n=============================================`);
  console.log(`[${dbName}] Memulai pembersihan...`);
  console.log(`[${dbName}] Ukuran awal: ${initialSize.toFixed(2)} MB (${(initialSize / 1024).toFixed(2)} GB)`);

  const dbUrl = `file:${dbPath}`;
  const db = createClient({ url: dbUrl });

  try {
    // 1. Hitung jumlah log sebelum dihapus
    const countBeforeRes = await db.execute(`SELECT COUNT(*) as count FROM activity_logs`);
    const countBefore = Number((countBeforeRes.rows[0] as any).count);
    console.log(`[${dbName}] Jumlah activity_logs saat ini: ${countBefore.toLocaleString('id-ID')} baris`);

    // 2. Hapus log yang lebih tua dari 7 hari
    console.log(`[${dbName}] Menghapus log yang lebih tua dari ${DAYS_TO_KEEP} hari...`);
    const deleteRes = await db.execute({
      sql: `DELETE FROM activity_logs WHERE created_at < date('now', ?)`,
      args: [`-${DAYS_TO_KEEP} days`]
    });
    
    const countAfterRes = await db.execute(`SELECT COUNT(*) as count FROM activity_logs`);
    const countAfter = Number((countAfterRes.rows[0] as any).count);
    const deletedCount = countBefore - countAfter;

    console.log(`[${dbName}] Berhasil menghapus: ${deletedCount.toLocaleString('id-ID')} baris`);
    console.log(`[${dbName}] Sisa activity_logs: ${countAfter.toLocaleString('id-ID')} baris`);

    // 3. Jalankan VACUUM untuk mengecilkan file di disk
    console.log(`[${dbName}] Menjalankan VACUUM... (Proses ini memakan waktu beberapa menit, mohon tunggu)`);
    const startTime = Date.now();
    await db.execute(`VACUUM`);
    const duration = ((Date.now() - startTime) / 1000).toFixed(1);
    console.log(`[${dbName}] VACUUM selesai dalam ${duration} detik.`);

    // 4. Hitung ukuran akhir
    const finalSize = getFileSizeInMB(dbPath);
    const savedSize = initialSize - finalSize;
    console.log(`[${dbName}] Ukuran akhir: ${finalSize.toFixed(2)} MB (${(finalSize / 1024).toFixed(2)} GB)`);
    console.log(`[${dbName}] Menghemat ruang disk: ${savedSize.toFixed(2)} MB`);

  } catch (error: any) {
    console.error(`[${dbName}] Terjadi kesalahan:`, error.message);
  } finally {
    db.close();
  }
}

async function main() {
  console.log(`=== SINTAK DATABASE CLEANUP & VACUUM TOOL ===`);
  console.log(`Menyimpan log untuk ${DAYS_TO_KEEP} hari terakhir.`);
  
  await cleanupDatabase('database_dev.sqlite');
  await cleanupDatabase('database.sqlite');
  
  console.log(`\n=============================================`);
  console.log(`Semua proses pembersihan selesai.`);
}

main().catch(err => {
  console.error("Terjadi kesalahan utama:", err);
});
